gossipsub.go

```diff
@@ -146,6 +146,30 @@ func WithPeerExchange(doPX bool) Option {
 	}
 }
 
+// WithDirectPeers is a gossipsub router option that specifies peers with direct
+// peering agreements. These peers are connected outside of the mesh, with all (valid)
+// message unconditionally forwarded to them. The router will maintain open connections
+// to these peers. Note that the peering agreement should be reciprocal with direct peers
+// symmetrically configured at both ends.
+func WithDirectPeers(pis []peer.AddrInfo) Option {
+	return func(ps *PubSub) error {
+		gs, ok := ps.rt.(*GossipSubRouter)
+		if !ok {
+			return fmt.Errorf("pubsub router is not gossipsub")
+		}
+
+		direct := make(map[peer.ID]struct{})
+		for _, pi := range pis {
+			direct[pi.ID] = struct{}{}
+			ps.host.Peerstore().AddAddrs(pi.ID, pi.Addrs, peerstore.PermanentAddrTTL)
+		}
+
+		gs.direct = direct
+
+		return nil
+	}
+}
```

Comment:
The kinds of users using this option are likely to form a federated/hierarchical network topology. In the case of Filecoin, there are probably two types of links:

* intra-miner: miners will probably operate N network peers for redundancy. They will want to tag each of their peers as direct peers. Effectively this will create a flooding behaviour _inside_ their network.
* inter-miner: peering miners will not want to point to _all_ network peers of their counterparts; that would defeat the purpose of gossipsub and would end up devolving into an inefficient flooding network. Approaches:
  * each miner nominates a subset of all their peers as their "externally interfacing peers", and other peers add only those peers in their peering agreements. This might be too static.
  * introduce "peering groups" -- each group lists all the peers of the counterpart, with a _target active peer count_, e.g. 2 or 3. We would take care of peering with at most that many peers from that group.

Reply:
These are interesting considerations. For now, @jbenet just wants to have direct peering agreement support without much further complexity.

Reply:
We can revisit when the need for peering groups arises.